src/v/datalake/coordinator/catalog_factory.cc

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+#include "config/configuration.h"
+#include "datalake/logger.h"
+#include "iceberg/catalog.h"
+#include "iceberg/filesystem_catalog.h"
+#include "iceberg/rest_catalog.h"
+
+namespace datalake::coordinator {
+
+namespace {
+
+// TODO: make this configurable.
+const ss::sstring base_location = "redpanda-iceberg-catalog";
+
+} // namespace
+
+std::unique_ptr<iceberg::catalog> create_catalog(
+  cloud_io::remote& io,
+  const cloud_storage_clients::bucket_name& bucket,
+  const config::configuration&) {
+    vlog(
+      datalake_log.info,
+      "Creating filesystem catalog with bucket: {} and location: {}",
+      bucket,
+      base_location);
+    return static_cast<std::unique_ptr<iceberg::catalog>>(
```

Comment:
you shouldn't need this cast here. you can directly return std::make_unique<Derived>(...);